chore: add project creation limit enforcement and tests

zlwaterfield opened the pull request:
## Changes

Enforces a hard limit of 1500 non-demo projects per organization and improves logic for project creation limits based on feature flags. Adds comprehensive tests to verify project creation behavior under various feature and limit scenarios, including demo project handling and update permissions.

## How did you test this code?

Added tests - also added some others for the `PremiumMultiProjectPermission` class.

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

zlwaterfield committed:
Add project creation limit enforcement and tests

Enforces a hard limit of 1000 non-demo projects per organization and improves logic for project creation limits based on feature flags. Adds comprehensive tests to verify project creation behavior under various feature and limit scenarios, including demo project handling and update permissions.

yasen-posthog commented on posthog/api/project.py:
```diff
@@ -817,12 +817,18 @@ def has_permission(self, request: request.Request, view) -> bool:
             if organization.teams.filter(is_demo=True).count() > 0:
                 return False
 
+        MAX_ALLOWED_PROJECTS_PER_ORG = 1000
```
+1

pawel-cebula reviewed (changes requested):
1. This change should be combined with updates to the `billing` YML files and fixtures, where it's currently still unlimited (+ `posthog` fixtures)
2. The error message should be updated as it currently states `You must upgrade your PostHog plan to be able to create and manage more projects.`, which is not true in the case of reaching the limit of 1,000 projects. We could modify it in those cases to say sth like `You've reached the limit of maximum 1,000 projects per organization. If you need more projects, contact us at ?@posthog.com.`
3. If we merge this, anyone who already has 1,000 projects
	- Won't be able to create new projects - so they'll start getting errors, not sure if it would impact their user's experience on production - so might be worth giving them heads up, e.g. we're going to merge this in 7 days
	- Won't be able to update existing projects (see `CREATE_ACTIONS = ["create", "update"]`) - I assume we only want this check for `create`
4. I see we have similar/same code in `posthog/api/team.py`, which I think is related to environments - do we need to implement same changes there too or no need anymore?

zlwaterfield committed:
Update limit error messages for projects and environments

Replaces generic upgrade messages with more specific error messages indicating the maximum allowed projects or environments for the current plan. Also introduces a constant for the hard project limit and ensures the message reflects the actual limit when exceeded. Updates related tests to match the new error messages.

zlwaterfield committed:
Update error message

zlwaterfield committed:
Update max limit 1,500

patricio-posthog reviewed (approved):
I agree no need to update the Billing public page, since this limit is more artificial due to scalability, not something we've decided based on pricing/features/etc.

Code looks good to me. Only concern could be race condition, I can see someone firing many requests at the same time and creating more than 1,000 projects (I think we received a similar security ticket lately). But I think it isn't a concern for now.

zlwaterfield committed:
Update project hard limit test to 1500 projects

Changed the test for project creation hard limit to use 1500 projects instead of 1000, updating mock values and error messages accordingly. This ensures the test reflects the new organization project limit.

pawel-cebula reviewed (approved):
Looks good, just one copy issue in the error message, approving in advance

zlwaterfield committed:
Update project limit message for clarity

Revised the message shown when the maximum project limit is reached to remove redundant wording. Updated related test to match the new message.
